server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java

```diff
@@ -21,48 +21,147 @@
+    private final String dataPathPrefix;
+    private final SimpleDateFormat dataPathSuffixFormat = new SimpleDateFormat("yyMMddHHmmss.SSS");
+    private final ParquetConfig parquetConfig;
 
     private TableSchema schema = null;
 
+    @Getter
+    private long firstWalLogId = -1;
+
+    private long lastWalLogId = -1;
+
+    @Getter
+    private long lastUpdateTime = 0;
+
     private final TreeMap<Object, Map<String, Object>> recordMap = new TreeMap<>();
 
     private final Lock lock = new ReentrantLock();
 
-    public MemoryTableImpl(String tableName, WalManager walManager) {
+    public MemoryTableImpl(String tableName,
+            WalManager walManager,
+            StorageAccessService storageAccessService,
+            String dataRootPath,
+            ParquetConfig parquetConfig) {
         this.tableName = tableName;
         this.walManager = walManager;
+        this.storageAccessService = storageAccessService;
+        this.parquetConfig = parquetConfig;
+        var path = dataRootPath;
+        if (!path.isEmpty() && !path.endsWith("/")) {
+            path += "/";
+        }
+        this.dataPathPrefix = path + "snapshots/" + tableName + "_._v0_";
+        this.dataPathSuffixFormat.setTimeZone(TimeZone.getTimeZone("GMT"));
+        this.load();
+    }
+
+    private void load() {
+        try {
+            var path = this.storageAccessService.list(dataPathPrefix).max(Comparator.naturalOrder()).orElse(null);
+            if (path == null) {
+                return;
+            }
+            var conf = new Configuration();
+            try (var reader = new SwParquetReaderBuilder(this.storageAccessService, path).withConf(conf).build()) {
+                for (; ; ) {
+                    var record = reader.read();
+                    if (record == null) {
+                        break;
+                    }
+                    if (this.schema == null) {
+                        this.schema = TableSchema.fromJsonString(conf.get(SwReadSupport.SCHEMA_KEY));
+                        var metadata = new ObjectMapper().readValue(conf.get(SwReadSupport.META_DATA_KEY), Map.class);
+                        this.lastWalLogId = Long.parseLong((String) metadata.get(LOG_ID_KEY));
+                        this.lastUpdateTime = Long.parseLong((String) metadata.get(LAST_UPDATE_TIME_KEY));
+                    }
+                    this.recordMap.put(record.get(this.schema.getKeyColumn()), record);
+                }
+            }
+        } catch (IOException e) {
+            throw new SwProcessException(ErrorType.DATASTORE, "failed to load " + this.tableName, e);
+        }
     }
 
+    @Override
+    public void save() {
+        String metadata;
+        try {
+            metadata = new ObjectMapper().writeValueAsString(
+                    Map.of(LOG_ID_KEY, Long.toString(this.lastWalLogId),
+                            LAST_UPDATE_TIME_KEY, Long.toString(this.lastUpdateTime)));
```

Comment:
Would it be better if `metadata` was maintained by protobuf?

Reply:
Done